### PR TITLE
Changed World, Agent and the Interpolation Model

### PR DESCRIPTION
**Pull request by patrickhart**
Behavior models can now have an arbitrary planning horizon and time discretization.
The Execute function has been moved from the agent to the world.
Additionally, execution models now also can set a valid or invalid flag.

An assert has been introduced in the World::Execute(..) function to make sure all agents have the same world time.

### CODE REVIEW

File: docs/source/models.md [hunk 1]
```diff
@@ -28,7 +28,7 @@ class BehaviorModel : public modules::commons::BaseType {
         last_action_(),
         behavior_status_(status) {}
   ...
-  virtual Trajectory Plan(float delta_time,
+  virtual Trajectory Plan(float min_dt,
```
Comment (juloberno) [inline docs/source/models.md:31]: may be call min_planning_time ?

File: modules/models/behavior/behavior_model.hpp [hunk 2]
```diff
@@ -73,19 +77,26 @@ class BehaviorModel : public modules::commons::BaseType {
...
+  Action GetAction() const { return action_to_behavior_; }
+  virtual void ActionToBehavior(const Action& action) {
+    action_to_behavior_ = action;
+  };
 
  private:
   dynamic::Trajectory last_trajectory_;
+  // can either be the last action or action to be executed
   Action last_action_;
+  Action action_to_behavior_;
```
Comment (juloberno) [inline modules/models/behavior/behavior_model.hpp:99]: Do we need to support the action to behavior for all behavior models? I think it is sufficient for motion primitives or these types of behavior models

File: modules/models/execution/interpolation/interpolate.cpp [hunk 3]
```diff
@@ -13,16 +13,115 @@ namespace modules {
...
+    if (world_time >= trajectory(i, TIME_POSITION)) {
+      ret_idx = i;
+      found_closest_pt = true;
+      break;
+    }
+  }
+  return {ret_idx, found_closest_pt};
+}
+
+State ExecutionModelInterpolate::Interpoalte(
```
Comment (juloberno) [inline modules/models/execution/interpolation/interpolate.cpp:75]: typo

File: modules/models/execution/execution_model.hpp [hunk 4]
```diff
@@ -22,33 +22,58 @@ using dynamic::DynamicModelPtr;
...
   ExecutionModel(const ExecutionModel& execution_model) :
     BaseType(execution_model.GetParams()),
-    last_trajectory_(execution_model.GetLastTrajectory()) {}
+    last_state_(execution_model.GetLastState()),
+    last_trajectory_(execution_model.GetLastTrajectory()),
+    execution_status_(execution_model.GetExecutionStatus()) {}
 
   virtual ~ExecutionModel() {}
 
+  State GetLastState() const { return last_state_; }
```
Comment (juloberno) [inline modules/models/execution/execution_model.hpp:48]: maybe call this GetExecutedState() ? last state could be unclear

File: modules/world/world.cpp [hunk 5]
```diff
@@ -41,6 +43,50 @@ World::World(const std::shared_ptr<World>& world) :
...
+void World::Execute(const float& world_time) {
+  using models::dynamic::StateDefinition::TIME_POSITION;
+  for (auto agent : agents_) {
+    if (agent.second->GetBehaviorStatus() == BehaviorStatus::VALID &&
+        agent.second->GetExecutionStatus() == ExecutionStatus::VALID) {
+      agent.second->UpdateStateAction();
+      // make sure all agents have the same world time
+      // otherwise the simulation is not correct
+      const auto& agent_state = agent.second->GetCurrentState();
+      assert(fabs(agent_state(TIME_POSITION) - world_time) < 0.01);
```
Comment (juloberno) [inline modules/world/world.cpp:74]: use BARK_EXPECT_TRUE here

File: modules/world/world.cpp [hunk 6]
```diff
@@ -41,6 +43,50 @@ World::World(const std::shared_ptr<World>& world) :
...
+  }
+  RemoveInvalidAgents();
+}
+
+WorldPtr World::GetWorldAtTime(const float& world_time) const {
+  WorldPtr current_world_state(this->Clone());
+  for (auto agent : current_world_state->GetAgents()) {
+    if (agent.second->GetBehaviorStatus() == BehaviorStatus::VALID)
+      agent.second->PlanExecution(world_time);
+    agent.second->UpdateStateAction();
```
Comment (juloberno) [inline modules/world/world.cpp:85]: if Execution Valid should be checked here

File: modules/models/behavior/behavior_model.hpp (hunk 2 above)
Comment (patrickhart) [reply modules/models/behavior/behavior_model.hpp:99]: Due to the fact that is is wrapped in the PyBehaviorModel I think there is no other option?

File: docs/source/models.md (hunk 1 above)
Comment (patrickhart) [reply docs/source/models.md:31]: Will change that!

File: modules/models/execution/interpolation/interpolate.cpp (hunk 3 above)
Comment (patrickhart) [reply modules/models/execution/interpolation/interpolate.cpp:75]: Ups.

File: modules/models/execution/execution_model.hpp (hunk 4 above)
Comment (patrickhart) [reply modules/models/execution/execution_model.hpp:48]: sounds good!

File: modules/world/world.cpp (hunk 5 above)
Comment (patrickhart) [reply modules/world/world.cpp:74]: okay!

File: modules/world/world.cpp (hunk 6 above)
Comment (patrickhart) [reply modules/world/world.cpp:85]: I do not think so, as PlanExecution calls the exxecution model of the agent.

**Pull request merged**